### PR TITLE
Update SCI model to SCIsegV2 and standardize on `nnUNetTrainer` folder structure

### LINKED ISSUE #4582: Update SCIseg model to the latest version
With [SCIsegV2 preprint](https://arxiv.org/pdf/2407.17265) out and the [model released](https://github.com/ivadomed/model_seg_sci/releases/tag/r20240729), update the SCI model to the latest version

### PR DESCRIPTION
**Pull request by naga-karthik**
#### GitHub

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [ ] I've applied [the relevant labels](https://intranet.neuro.polymtl.ca/onboarding/software-development.html#pr-labels) to this PR
- [x] I've assigned a reviewer

## Description

This tiny PR updates our SCI model to its latest version: SCIsegV2. I have also update the model description by adding a few details about the training dataset. 

Resolves #4582, https://github.com/ivadomed/model_seg_sci/issues/89

### CODE REVIEW

File: spinalcordtoolbox/deepseg/models.py [hunk 1]
```diff
@@ -131,9 +131,9 @@
     },
     "model_seg_sci_multiclass_sc_lesion_nnunet": {
         "url": [
-            "https://github.com/ivadomed/model_seg_sci/releases/download/r20240130/model-sci-multisite_r20240130.zip"
+            "https://github.com/ivadomed/model_seg_sci/releases/download/r20240729/model_SCIsegV2_r20240729.zip"
         ],
-        "description": "Traumatic SCI spinal cord/lesion segmentation for T2w contrast",
+        "description": "Universal tool for intramedullary spinal cord and lesion segmentation in T2w MRI",
```
Comment (valosekj) [inline spinalcordtoolbox/deepseg/models.py:136]:
```suggestion
        "description": "Universal tool for segmentation of intramedullary lesion and spinal cord in T2w MRI",
```

File: spinalcordtoolbox/deepseg/models.py [hunk 2]
```diff
@@ -293,11 +293,12 @@
          'url': 'https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/',
          'models': ['model_seg_sc_contrast_agnostic_softseg_monai']},
     'seg_sc_lesion_t2w_sci':
-        {'description': 'Traumatic SCI spinal cord/lesion seg for T2w contrast',
+        {'description': 'Universal tool for intrameduallary spinal cord and lesion segmentation in T2w MRI',
```
Comment (valosekj) [inline spinalcordtoolbox/deepseg/models.py:296]:
```suggestion
        {'description': 'Universal tool for segmentation of intramedullary lesion and spinal cord in T2w MRI',
```

File: spinalcordtoolbox/deepseg/models.py [hunk 3]
```diff
@@ -131,9 +131,9 @@
     },
     "model_seg_sci_multiclass_sc_lesion_nnunet": {
         "url": [
-            "https://github.com/ivadomed/model_seg_sci/releases/download/r20240130/model-sci-multisite_r20240130.zip"
+            "https://github.com/ivadomed/model_seg_sci/releases/download/r20240729/model_SCIsegV2_r20240729.zip"
```
Comment (valosekj) [inline spinalcordtoolbox/deepseg/models.py:134]: Note:
- for this new release [r20240729](https://github.com/ivadomed/model_seg_sci/releases/tag/r20240729), the nnUNet output structure is kept intact, thus allowing it to be used in the 3DSlicer software as well (as discussed in https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4531#issuecomment-2181169132).
- while for the previous release [r20240130](https://github.com/ivadomed/model_seg_sci/releases/tag/r20240130), the nested folders were removed

I'm wondering whether the `sct_deepseg` code is already compatible with the nested nnUNet folder structure

Comment (naga-karthik) [reply spinalcordtoolbox/deepseg/models.py:134]: Aha! You were right! the nested structure is not supported by SCT

<details>
<summary> Error trace </summary>

```python
Removing temporary folders...
Traceback (most recent call last):
  File "/home/GRAMES.POLYMTL.CA/u114716/spinalcordtoolbox/spinalcordtoolbox/scripts/sct_deepseg.py", line 380, in <module>
    main(sys.argv[1:])
  File "/home/GRAMES.POLYMTL.CA/u114716/spinalcordtoolbox/spinalcordtoolbox/scripts/sct_deepseg.py", line 289, in main
    im_lst, target_lst = inference.segment_non_ivadomed(path_model, model_type, input_filenames, thr,
  File "/home/GRAMES.POLYMTL.CA/u114716/spinalcordtoolbox/spinalcordtoolbox/deepseg/inference.py", line 121, in segment_non_ivadomed
    net = create_net(path_model, device)
  File "/home/GRAMES.POLYMTL.CA/u114716/spinalcordtoolbox/spinalcordtoolbox/deepseg/nnunet.py", line 23, in create_nnunet_from_plans
    raise FileNotFoundError(f"No 'fold_*' directories found in model path: {path_model}")
FileNotFoundError: No 'fold_*' directories found in model path: /home/GRAMES.POLYMTL.CA/u114716/spinalcordtoolbox/data/deepseg_models/model_seg_sci_multiclass_sc_lesion_nnunet     
```

</details>

Comment (naga-karthik) [reply spinalcordtoolbox/deepseg/models.py:134]: updated to handle nested structure here https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4583/commits/4db6ac3c69d15fe58d532634991ec514631f1f6f

Comment (valosekj) [reply spinalcordtoolbox/deepseg/models.py:134]: Great! Thank you!

File: spinalcordtoolbox/deepseg/models.py [hunk 4]
```diff
@@ -293,11 +293,12 @@
          'url': 'https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/',
          'models': ['model_seg_sc_contrast_agnostic_softseg_monai']},
     'seg_sc_lesion_t2w_sci':
-        {'description': 'Traumatic SCI spinal cord/lesion seg for T2w contrast',
```
Comment (joshuacwnewton) [inline spinalcordtoolbox/deepseg/models.py:296]: Note that this description is displayed when running `sct_deepseg -h`, so it might be worth keeping it short (to avoid wrapping in the table):

image

File: spinalcordtoolbox/deepseg/nnunet.py [hunk 5]
```diff
@@ -18,6 +18,13 @@
 
 def create_nnunet_from_plans(path_model, device: 'torch.device'):
     tile_step_size = 0.5
+    # get the nnunet trainer directory
+    trainer = os.listdir(path_model)[0]
+    if not trainer.startswith("nnUNetTrainer"):
+        raise FileNotFoundError(f"Could not find 'nnUNetTrainer*' directory inside model path: {path_model} "
+                                "Please make sure the release keeps the nnUNet output structure intact "
+                                "by also including the 'nnUNetTrainer*' directory.")
+    path_model = os.path.join(path_model, trainer)
```
Comment (joshuacwnewton) [inline spinalcordtoolbox/deepseg/nnunet.py:27]: It looks like this check is failing the other non-SCIseg nnunet models. 

For example, the rootlets model does not contain `nnUNetTrainer` at all:

image

I think that, instead of looking for nnUNetTrainer, it is better to look for the presence of the `.pth` file in subfolders:

image

This is similar to what the code already does when checking if a model folder is valid:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/6fe73ed295b61f37bdbbc4c2914f576f7c6ca2c6/spinalcordtoolbox/deepseg/models.py#L470-L478

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/6fe73ed295b61f37bdbbc4c2914f576f7c6ca2c6/spinalcordtoolbox/deepseg/models.py#L503-L507

Comment (valosekj) [reply spinalcordtoolbox/deepseg/nnunet.py:27]: > For example, the rootlets model does not contain nnUNetTrainer at all:

Thanks for catching this! I have an update of the rootlets release on my to-do list.

Comment (naga-karthik) [reply spinalcordtoolbox/deepseg/nnunet.py:27]: Yes, I believe it does! Nilser also posted the folder hierarchy and we see that there's a `nnUNetTrainer*` folder

```bash
Dataset403_seg_ms_lesion_mp2rage_1mm_322subj/
└── nnUNetTrainer_seg_ms_lesion_mp2rage_nnUNetPlans_3d_fullres
    ├── dataset_fingerprint.json
    ├── dataset.json
    ├── fold_3
    │   ├── checkpoint_final.pth
    │   ├── debug.json
    │   ├── progress.png
    │   └── training_log_2024_3_14_20_35_08.txt
    └── plans.json 
```

Comment (valosekj) [reply spinalcordtoolbox/deepseg/nnunet.py:27]: Just to confirm, I plan to use the following folder hierarchy for the rootlets model:

```console
model-spinal-rootlets_M5_r20240129-2
└── nnUNetTrainer_2000epochs__nnUNetPlans__3d_fullres
    ├── dataset.json
    ├── dataset_fingerprint.json
    ├── fold_all
    │   ├── checkpoint_final.pth
    │   ├── debug.json
    │   ├── progress.png
    │   ├── training_log_2024_1_9_15_28_27.png
    │   └── training_log_2024_1_9_15_28_27.txt
    └── plans.json
```

instead of the current:

```console
model-spinal-rootlets_M5_r20240129
├── dataset.json
├── dataset_fingerprint.json
├── fold_all
│   ├── checkpoint_final.pth
│   ├── debug.json
│   ├── progress.png
│   ├── training_log_2024_1_9_15_28_27.png
│   └── training_log_2024_1_9_15_28_27.txt
└── plans.json
```

File: spinalcordtoolbox/deepseg/models.py (hunk 4 above)
Comment (naga-karthik) [reply spinalcordtoolbox/deepseg/models.py:296]: Done in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4583/commits/5cdc2e6f25ed78ce5cf40e88173413fc22fe014f, I believe it should be fine now?

File: spinalcordtoolbox/deepseg/nnunet.py (hunk 5 above)
Comment (valosekj) [reply spinalcordtoolbox/deepseg/nnunet.py:27]: The rootlets model updated in: https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4583/commits/cd66b0717436a3b5f5fd95de294a1c034c86f3e6

Comment (joshuacwnewton) [reply spinalcordtoolbox/deepseg/nnunet.py]: The check for `nnUNetTrainer` was failing because it was finding `source.json` instead of the `nnUNetTrainer` folder (oops!) so I fixed the check.

Now, the tests are failing for a different reason:

- [ ] `seg_ms_lesion_mp2rage` does not have the nnUNetTrainer folder structure.
- [ ] Dice score for new SCISeg model is less than expected (`0.8964488212473889`, should be `> 0.95`)
   - Note that this is for a fake lesion, so it isn't necessarily representative of model performance. But, here is the comparison of the newly-segmented fake lesion vs. the ground truth (blue is gt from old model, red is new model): 

ezgif-1-d8d48c7a63

To me, this seems like an improvement? (I remember that the old model had trouble detecting the intense region on the cropped image, so I had to manually add a VERY intense region. But, now, the new model properly detects both the natural and synthetic intense regions.)

So, I think it should be fine to just lower the dice requirement for now to make the test pass? (>0.85)

Comment (naga-karthik) [reply spinalcordtoolbox/deepseg/nnunet.py]: Thanks for the testing, Joshua! This is rather concerning, the new model (i.e. red mask) is kind of oversegmenting -- in this case, it's probably also segmenting the spinal canal in addition to the lesion (which we don't want)

Could you please also try the `-keep_largest` postprocessing to see if we have something better?

Comment (joshuacwnewton) [reply spinalcordtoolbox/deepseg/nnunet.py]: > Could you please also try the `-keep_largest` postprocessing to see if we have something better?

`-largest 1` does fix the lesion segmentation! However, it turns out that the `0.89` dice score was actually from the SC seg. (Meaning, even though we have an improvement in the lesion seg, the test still fails for the SC).

Here is the SC comparison:

ezgif-1-abeede9c05

So, it seems like our GT (old model) under-segments the SC, while the new model segments it better. Meaning, the test failure is definitely an issue with our ground truth (I think).

Comment (naga-karthik) [reply spinalcordtoolbox/deepseg/nnunet.py]: oh for sure, the blue GT mask from the old model is not good enough. Yeah, we should update our GT mask or lower the threshold for the tests

Comment (valosekj) [reply spinalcordtoolbox/deepseg/nnunet.py]: > Yeah, we should update our GT mask 

lmk if you need help with this, I'm happy to help

Comment (joshuacwnewton) [reply spinalcordtoolbox/deepseg/nnunet.py]: > [ ] `seg_ms_lesion_mp2rage` does not have the nnUNetTrainer folder structure.

Just to follow up on this -- It turns out that the MP2RAGE model _does_ have the nnUNetTrainer structure:

image

The issue, though, is that `sct_deepseg -install` removes the parent folder, but _only if there are no other files apart from the 1 subfolder_:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/9cd5700c7e05baf7ebd5fd56e74e8779029f528f/spinalcordtoolbox/download.py#L307-L313

image

The reason this doesn't happen for the other models is purely coincidence... there is a `.DS_Store` file in the .zip! :facepalm:

I have pushed a fix that will preserve `nnUNetTrainer` subdirectories.

Comment (valosekj) [reply spinalcordtoolbox/deepseg/nnunet.py]: Goot catch, Joshua! Thank you for fixing the MP2RAGE model 🙏🏻

File: spinalcordtoolbox/download.py [hunk 6]
```diff
@@ -310,7 +310,12 @@ def install_data(url, dest_folder, keep=False):
             for entry in it:
                 if entry.name in ("__MACOSX",):
                     continue
-                bundle_folder = entry.path
+                # NB: Don't traverse into nnUNetTrainer folders, and instead preserve the structure.
+                # This allows users to re-use the model folder for 3DSlicer with the correct structure.
+                elif entry.name.startswith("nnUNetTrainer"):
+                    bundle_folder = extraction_folder
+                else:
+                    bundle_folder = entry.path
```
Comment (joshuacwnewton) [inline spinalcordtoolbox/download.py:318]: I don't love having a reference to `nnUNetTrainer` inside of `download.py`, which is more "general"... :thinking:

File: testing/cli/test_sct_deepseg.py [hunk 7]
```diff
@@ -174,4 +175,4 @@ def test_segment_nifti_multiclass(fname_image, fnames_seg_manual, fname_out, suf
             im_seg = Image(fname_out)
             im_seg_manual = Image(fname_seg_manual)
             dice_segmentation = compute_dice(im_seg, im_seg_manual, mode='3d', zboundaries=False)
-            assert dice_segmentation > 0.95
+            assert dice_segmentation > 0.85  # Temporary until we improve "fake_lesion" ground truth
```
Comment (joshuacwnewton) [inline testing/cli/test_sct_deepseg.py:178]: Todo: 

- [ ] Replace the ground truth file inside of `sct_testing_data`
- [ ] Create a new release
- [ ] Update the URL inside this PR
- [ ] Revert the change to the target dice score

**Review by joshuacwnewton — APPROVED** [review]
I think as soon as these two comments are addressed, this will be good to merge.

I can take care of these changes then merge? :thinking:

File: testing/cli/test_sct_deepseg.py (hunk 7 above)
Comment (joshuacwnewton) [reply testing/cli/test_sct_deepseg.py]: Done in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4583/commits/9ed7a4d09c2861cb4268b5fa8029350f4d270cfd.

File: spinalcordtoolbox/download.py (hunk 6 above)
Comment (joshuacwnewton) [reply spinalcordtoolbox/download.py]: Done in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4583/commits/913401180c9f6210df0cdaba73a33d423b27cec7.
